pkg/agent/manager/cache/cache.go

```diff
@@ -2,357 +2,582 @@ package cache
+		}
 	}
-	return &cacheImpl{
-		cache:       make(map[string]*Entry),
-		log:         log.WithField("subsystem_name", "cache"),
-		trustDomain: trustDomain,
-		bundles:     observer.NewProperty(bundles),
-		subscribers: NewSubscribers(),
-		jwtSVIDS:    make(map[string]*client.JWTSVID),
-	}
-}
-
-func (c *cacheImpl) SetBundles(newBundles map[string]*Bundle) {
-	// SetBundles() and Bundle()/Bundles() can be called concurrently since
-	// the "property" is atomic. Before the following code can merge in changes
-	// it needs to make a copy of the map to mutate so it doesn't modify
-	// the bundle map out from underneath readers. SetBundles() is not intended
-	// to be called by more than one goroutine at a time.
-
-	// copy the map
-	bundles := make(map[string]*Bundle)
-	for k, v := range c.Bundles() {
-		bundles[k] = v
-	}
-
-	// merge in changes
-	changed := false
-	for id, newBundle := range newBundles {
-		bundle := bundles[id]
-		if bundle == nil || !bundle.EqualTo(newBundle) {
-			bundles[id] = newBundle
-			changed = true
+
+	// add/update registration entries
+	for _, regEntry := range update.RegistrationEntries {
+		entry, entryExists := c.entries[regEntry.EntryId]
+		if !entryExists {
+			entry = newCacheEntry()
+			c.entries[regEntry.EntryId] = entry
+		}
+
+		selAdd, selRem := c.diffSelectors(set, entry, entry.regEntry, regEntry)
+		fedAdd, fedRem := c.diffFederatesWith(entry.regEntry, regEntry)
+
+		entry.regEntry = regEntry
+
+		// Now figure out if something related to the entry has changed so
+		// interested subscribers can be notified.
+		federatedBundlesChanged := fedAdd > 0 || fedRem > 0
+		if !federatedBundlesChanged {
+			for _, id := range regEntry.FederatesWith {
+				if bundleChanged[id] {
+					federatedBundlesChanged = true
+					break
+				}
+			}
+		}
+
+		// Check if the X509-SVID was updated for this entry
+		svid, svidUpdated := update.X509SVIDs[regEntry.EntryId]
+		if svidUpdated {
+			entry.svid = svid
+		}
+
+		metaUpdated := federatedBundlesChanged || svidUpdated
+
+		// something related to this entry changed outside of the selectors.
+		// make sure subscribers are notified.
+		if metaUpdated {
+			set.Merge(regEntry.Selectors...)
+		}
+
+		// Invoke the svid checker callback for this entry
+		if checkSVID != nil {
+			checkSVID(regEntry, entry.svid)
+		}
+
+		if selAdd > 0 || selRem > 0 || fedAdd > 0 || fedRem > 0 || svidUpdated {
+			log := c.log.WithFields(logrus.Fields{
+				"entry_id":  regEntry.EntryId,
+				"spiffe_id": regEntry.SpiffeId,
+			})
+			if svidUpdated {
+				log = log.WithField("svid_updated", svidUpdated)
+			}
+			if selAdd > 0 {
+				log = log.WithField("sel_add", selAdd)
+			}
+			if selRem > 0 {
+				log = log.WithField("sel_rem", selRem)
+			}
+			if fedAdd > 0 {
+				log = log.WithField("fed_add", fedAdd)
+			}
+			if fedRem > 0 {
+				log = log.WithField("fed_rem", fedRem)
+			}
+			if entryExists {
+				log.Debug("Entry updated")
+			} else {
+				log.Debug("Entry created")
```

Comment:
I added it because this refactor was huge and I wanted to log as much info as possible in case there were bugs introduced. It doesn't print the selectors or bundles themselves, but rather a count of the changes. Something like:

```
DEBU[0000] Entry created                                 entry_id=0a4eaff0-1ad8-4624-8d74-93c07fb1c9e6 sel_add=1 spiffe_id="spiffe://example.org/node" subsystem_name=cache
INFO[0000] Renewing X509-SVID                            spiffe_id="spiffe://example.org/node" subsystem_name=manager
DEBU[0000] Entry updated                                 entry_id=0a4eaff0-1ad8-4624-8d74-93c07fb1c9e6 spiffe_id="spiffe://example.org/node" subsystem_name=cache svid_updated=true
```

I'd like to keep it (or at least some version of it). If you are uncomfortable with the verbosity, i could log the ugliness under Trace level or something.... thoughts?

Reply:
Fair enough. There is no trace level IIRC so debug is fine